cylc/flow/id_match.py

```diff
@@ -35,54 +33,18 @@
 from cylc.flow.cycling.loader import get_point
 
 if TYPE_CHECKING:
-    # from typing_extensions import Literal
-
     from cylc.flow.task_pool import Pool
-    from cylc.flow.task_proxy import TaskProxy
+    from cylc.flow.task_hold_mgr import PseudoTaskProxy
     from cylc.flow.cycling import PointBase
 
 
-# @overload
-# def filter_ids(
-#     pool: 'Pool',
-#     ids: 'Iterable[str]',
-#     *,
-#     warn: 'bool' = True,
-#     out: 'Literal[IDTokens.Task]' = IDTokens.Task,
-#     pattern_match: 'bool' = True,
-# ) -> 'Tuple[List[TaskProxy], List[str]]':
-#     ...
-#
-#
-# @overload
-# def filter_ids(
-#     pool: 'Pool',
-#     ids: 'Iterable[str]',
-#     *,
-#     warn: 'bool' = True,
-#     out: 'Literal[IDTokens.Cycle]' = IDTokens.Cycle,
-#     pattern_match: 'bool' = True,
-# ) -> 'Tuple[List[PointBase], List[str]]':
-#     ...
-
-
-# _RET = (
-#     'Union['
-#     'Tuple[List[TaskProxy], List[str]]'
-#     ', '
-#     'Tuple[List[PointBase], List[str]]'
-#     ']'
-# )
-
-
 def filter_ids(
-    pool: 'Pool',
+    pool: Union['Pool', Dict['PointBase', Dict[str, 'PseudoTaskProxy']]],
```

Comment:
I'm not overly keen on letting these posers mingle with the real deal in the pool as they could get sucked through into other code where they don't belong causing runtime errors. Also could they interfere with spawning logic or cause flow merging?

Perhaps we could stick these into a special held pool?

Reply:
> I'm not overly keen on letting these posers mingle with the real deal in the pool

Firstly, I did that in response to your earlier review comment above 😁 :

> *If we reformatted the hold-list to match the format of the pool*, could we use the one method for both cases to avoid adding a new implementation?

The problem is `filter_ids` works on a task pool, whereas the hold list - because it also holds future tasks - has just `(task-name, cycle-point)` tuples.

Secondly, there's no "mixing" going on. The `filter_ids` method just iterates over the pool to extract information. It can return a list of tasks, the type of which depends on whether we're matching the active pool or future tasks, so the return value has to be treated accordingly. 

But the pseudo task proxies are only generated, for this specific purpose, by an explicit call to the task hold manager, so I don't see any reason to be concerned about them leaking into the real task pool somehow. And they can't interfere with spawning logic or anything else if they're not in the task pool.